src/raster/array.jl

```diff
@@ -1,89 +1,106 @@
+    iszero(nraster(ds)) && throw(ArgumentError("The Dataset does not contain any raster bands"))
+    s = _common_size(ds)
+    RasterDataset{_dataset_type(ds), typeof(ds)}(ds,s)
+end
+function _dataset_type(ds::AbstractDataset)
+  alldatatypes = map(1:nraster(ds)) do i
+    b = getband(ds,i)
+    pixeltype(b)
+  end
+  reduce(promote_type,alldatatypes)
+end
+function _common_size(ds::AbstractDataset)
+  nr = nraster(ds)
+  allsizes = map(1:nr) do i
+    b = getband(ds,i)
+    size(b)
+  end
+  s = unique(allsizes)
+  length(s) == 1 || throw(DimensionMismatch("Can not coerce bands to single dataset, different sizes found"))
+  Int64.((s[1]...,nr))
+end
+getband(ds::RasterDataset,i) = getband(ds.ds,i)
+unsafe_readraster(args...;kwargs...)  = RasterDataset(unsafe_read(args...;kwargs...))
+readraster(s::String;kwargs...) = RasterDataset(read(s;kwargs...))
+function eachchunk(ds::RasterDataset)
+  subchunks = eachchunk(getband(ds,1))
+  GridChunks(ds,(subchunks.chunksize...,1))
+end
+DiskArrays.haschunks(::RasterDataset) = DiskArrays.Chunked()
+DiskArrays.haschunks(::AbstractRasterBand) = DiskArrays.Chunked()
 # AbstractRasterBand indexing
 
 Base.size(band::AbstractRasterBand) = width(band), height(band)
-Base.firstindex(band::AbstractRasterBand, d) = 1
-Base.lastindex(band::AbstractRasterBand, d) = size(band)[d]
+# Don't know if we need these, commenting out for the moment
+#Base.firstindex(band::AbstractRasterBand, d) = 1
+#Base.lastindex(band::AbstractRasterBand, d) = size(band)[d]
 
-Base.getindex(band::AbstractRasterBand, x::AllowedXY, y::AllowedXY) = begin
-    I = Base.to_indices(band, (x, y))
-    buffer = Array{pixeltype(band),2}(undef, length.(I)...)
+function eachchunk(band::AbstractRasterBand)
+  wI = windows(band)
+  cs = wI.blockiter.xbsize,wI.blockiter.ybsize
+  GridChunks(band,cs)
+end
+
+DiskArrays.readblock!(band::AbstractRasterBand, buffer, x::AbstractUnitRange, y::AbstractUnitRange) = begin
     # Calculate `read!` args and read
-    xoffset, yoffset = first.(I) .- 1
-    xsize, ysize = length.(I)
+    xoffset, yoffset = first(x)-1, first(y)-1
+    xsize, ysize = length(x), length(y)
     read!(band, buffer, xoffset, yoffset, xsize, ysize)
-    # Drop dimensions of integer indices
-    dimsize = map(length, _dropint(I...))
-    if dimsize == ()
-        return buffer[1]
-    else
-        return reshape(buffer, dimsize...)
-    end
 end
 
-Base.setindex!(band::AbstractRasterBand, value, x::AllowedXY, y::AllowedXY) = begin
-    I = Base.to_indices(band, (x, y))
-    if value isa AbstractArray
-        Base.setindex_shape_check(value, length.(I)...)
-        value = reshape(value, length.(I))
-    else
-        value = reshape([value], 1, 1)
-    end
-    # Calculate `write!` args and write
-    xoffset, yoffset = first.(I) .- 1
-    xsize, ysize = length.(I)
+DiskArrays.writeblock!(band::AbstractRasterBand, value, x::AbstractUnitRange, y::AbstractUnitRange) = begin
+    # Calculate `read!` args and read
```

Comment:
I didn't understand this comment.

Reply:
Sorry, this was a copy-paste remnant from some other source.

Reply:
so these got removed now